enhancements/20240611-vm-schedule-backup.md

```diff
@@ -0,0 +1,195 @@
+
+The Harvester controller will suspend the schedule when the number of failed backups reaches a pre-defined (configurable) threshold. After making sure the backup target is normal, the user can ask to resume the scheduled backup, then the Harvester controller will check the remote target's health, and resume the scheduled job if certain conditions are met.
+
+## Proposal
+
+### User Stories
+
+Although [Longhorn provides recurring backup](https://longhorn.io/docs/1.6.2/snapshots-and-backups/scheduling-backups-and-snapshots/) that is per volume and not VM-based, so it brings the challenge to restore the VM from these volume backups.
+
+With this feature, the user can have a **list of VM backups/snapshots** which are kept renovated according to the format of a [CRON expression](https://en.wikipedia.org/wiki/Cron#CRON_expression), and the user also can choose anyone from the list to restore the VM.
+
+
+### API changes
+
+## Design
+
+### Implementation Overview
+
+#### CRD
+
+- Add new resource `ScheduleVMBackup` to control VM schedule backup jobs
+```
+type ScheduleVMBackupSpec struct {
+	// +kubebuilder:validation:Required
+	Cron string `json:"cron"`
+
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default:=8
+	// +kubebuilder:validation:Maximum=254
+	// +kubebuilder:validation:Minimum=2
+	Retain int `json:"retain"`
+
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default:=4
+	// +kubebuilder:validation:Minimum=2
+	MaxFailure int `json:"maxFailure"`
+
+	// +optional
+	// +kubebuilder:default:=false
+	Suspend bool `json:"suspend"`
+
+	// +kubebuilder:validation:Required
+	VMBackupSpec VirtualMachineBackupSpec `json:"vmbackup"`
+}
+```
+
+```
+type ScheduleVMBackupStatus struct {
+	// +optional
+	VMBackupInfo []VMBackupInfo `json:"vmbackupInfo,omitempty"`
+
+	// +optional
+	Failure int `json:"failure,omitempty"`
+
+	// +optional
+	Suspend bool `json:"suspend,omitempty"`
+
+	// +optional
+	Conditions []Condition `json:"conditions,omitempty"`
+}
+```
+- `spec.cron` specifies the cron expression for the VM backup schedule
+- `spec.retain` specifies the number of up-to-date VM backups to retain
+- `spec.maxFailure` specifies the max number of consecutive failed backups that could be tolerated, if the number of consecutive failed backups reaches this setting, the Harvester controller will suspend the schedule
+- `spec.suspend` is used to suspend/resume the schedule proactively
+- `spec.vmbackup` specifies the source VM to backup/snapshot
+- `status.vmbackupInfo` records the current backup/snapshot list, including the status of all the volume backups/snapshots for the VM
+- `status.failure` number of consecutive failure backups after the last successful backups
+- `status.suspend` presents if the schedule is suspended
```

Comment:
Shouldn't this named `suspended`? The `spec.suspend` describes an action (so present tense is correct), but `status.suspend` describes a status, so it needs to be named `status.suspended` (past tense).

Reply:
Updated per suggestion, will update the related implementation in the controller, too. Thanks.